op-program/host/host.go

```diff
@@ -80,54 +74,89 @@ func FaultProofProgram(logger log.Logger, cfg *config.Config) error {
 
 		logger.Info("Setting up pre-fetcher")
 		prefetch := prefetcher.NewPrefetcher(logger, l1Cl, l2DebugCl, kv)
-		preimageOracle = asOracleFn(func(key common.Hash) ([]byte, error) {
-			return prefetch.GetPreimage(ctx, key)
-		})
-		hinter = asHinter(prefetch.Hint)
+		getPreimage = func(key common.Hash) ([]byte, error) { return prefetch.GetPreimage(ctx, key) }
+		hinter = prefetch.Hint
 	} else {
 		logger.Info("Using offline mode. All required pre-images must be pre-populated.")
-		preimageOracle = asOracleFn(kv.Get)
-		hinter = func(v preimage.Hint) {
-			logger.Debug("ignoring prefetch hint", "hint", v)
+		getPreimage = kv.Get
+		hinter = func(hint string) error {
+			logger.Debug("ignoring prefetch hint", "hint", hint)
+			return nil
 		}
 	}
-	l1Source := l1.NewSource(logger, preimageOracle, hinter, cfg.L1Head)
 
-	l2Source, err := l2.NewEngine(logger, preimageOracle, hinter, cfg)
-	if err != nil {
-		return fmt.Errorf("connect l2 oracle: %w", err)
-	}
+	// Setup pipe for preimage oracle interaction
+	pClientRW, pHostRW := bidirectionalPipe()
+	oracleServer := preimage.NewOracleServer(pHostRW)
+	// Setup pipe for hint comms
+	hHostR, hClientW := io.Pipe()
+	hHost := preimage.NewHintReader(hHostR)
+	defer pHostRW.Close()
+	defer hHostR.Close()
+	routeHints(logger, hHost, hinter)
+	launchOracleServer(logger, oracleServer, getPreimage)
```

Comment:
This means that we have three threads going at once
1. `ClientProgram` (main thread)
2. `routeHints`
3. `oracleServer`
I think we should be able to do it in two - client program & host. The host will have to multiplex between the two different pipes but I think it should at least be spun up as one unit (maybe having two thread internally) to better coordinate between the two FDs

Reply:
It's doable but doing so complicates the host programming. The host oracle and prefetcher will have to be rewritten to support a poll-based interface. And doing so also exposes too much of the internals of pipes (since Go doesn't provide an idiomatic way to select on pipes).

Reply:
Ok. That makes sense for why we need three threads. There's another part here where I don't love spwaning them this way. One thing I think that would be interesting is spawning the client program in it's own routine/thread only after these routines & ready

Reply:
Yeah that's the end goal of this host-client I/O separation. Spawning the client program in its own thread will be done in a follow up PR